osu.Game/Rulesets/Mods/ModChallenge.cs

```diff
@@ -0,0 +1,98 @@
+
+        [SettingSource("Checking interval", "When conditions will be checked.")]
+        public Bindable<ChallengeCheckInterval> CheckingInterval { get; } = new Bindable<ChallengeCheckInterval>();
+
+        protected readonly List<HitObject> HitObjectsBeforeBreaks = new List<HitObject>();
+        protected HitObject LastHitObjectInBeatmap;
+
+        public void ApplyToBeatmap(IBeatmap beatmap)
+        {
+            switch (CheckingInterval.Value)
+            {
+                case ChallengeCheckInterval.AtEnd:
+                    LastHitObjectInBeatmap = beatmap.HitObjects.LastOrDefault();
+                    break;
+
+                case ChallengeCheckInterval.AtBreak:
+                    for (int i = 0; i < beatmap.HitObjects.Count - 1; i++)
+                    {
+                        double inBetweenTime = (beatmap.HitObjects.ElementAtOrDefault(i).GetEndTime() + beatmap.HitObjects.ElementAtOrDefault(i + 1).StartTime) / 2;
+
+                        foreach (BreakPeriod breakPeriod in beatmap.Breaks)
+                        {
+                            if (breakPeriod.Contains(inBetweenTime))
+                                HitObjectsBeforeBreaks.Add(beatmap.HitObjects.ElementAtOrDefault(i));
+                        }
+                    }
+
+                    break;
+            }
+        }
+
+        public virtual ScoreRank AdjustRank(ScoreRank rank, double accuracy) => rank;
+
+        /// <summary>
+        /// Check whether the <see cref="CheckingInterval"/> allows for a challenge-induced failure to take place at this <see cref="HitObject"/>.
+        /// </summary>
+        /// <returns>Whether a challenge-induced failure is allowed to take place.</returns>
+        protected bool AllowChallengeFailureAtHitObject(HitObject hitObject)
+        {
+            switch (CheckingInterval.Value)
+            {
+                case ChallengeCheckInterval.AtBreak:
+                    if (!HitObjectsBeforeBreaks.Contains(hitObject))
+                        return false;
+
+                    break;
+
+                case ChallengeCheckInterval.AtEnd:
+                    if (hitObject != LastHitObjectInBeatmap)
+                        return false;
+
+                    break;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Checks whether the <see cref="JudgementResult"/> is anything less than the best it can possibly be.
+        /// </summary>
+        /// <param name="judgement">The judgement to check.</param>
+        /// <returns>Whether the judgement is the best it can possibly be.</returns>
+        protected bool JudgementIsFlawed(JudgementResult judgement)
+        {
+            return !(judgement.Type == judgement.Judgement.MaxResult || judgement.Type == HitResult.LargeTickHit) && judgement.Type.AffectsCombo();
+        }
+
+        public enum ChallengeCheckInterval
+        {
+            Continuously,
+            AtBreak,
+            AtEnd,
+        }
```

Comment:
Should add descriptions, `AtBreak` is not good to read for the user. Should be `Continously`, `At break`, `At end`.